### ISSUE TITLE
Add -assoc-in and multi-key updates

### ISSUE THREAD

**Issue by lsenta** [replikativ/konserve#9]
I implemented a firebase store for konserve and the abstraction you provides is pretty handy: I can have multiple stores, memory, local, remote, etc and use whichever I need depending on my requirements.

But there are a few operations that I'd like to replace or implement:

- direct `assoc-in`:
  The `core/assoc-in` relies on `-update-in` which implies a costly network back & forth in remote 
  cases.
- multi-key `assoc-in`/`update-in`:
   Firebase allows batching multiple updates in a single, atomic, operation. The append operation does 3 updates that could be merged into a single one. We'd extends the `go-locked` system to lock over a set of key-vec.
   This would be backward compatible by providing a default implementation.

I understand that as a small, backward compatible change in the protocol.
Would you be open for a patch, and maybe give me a hand on getting the go-locked right?

**Comment by whilo** [comment]
Ok, we can reintroduce an -assoc-in protocol which can fall back to -update-in by default maybe... This is reasonable. I was a bit too minimalist there, I guess :)

Regarding multi-key. It is something in general that is on my radar, like having transactions over multiple keys in konserve in general. But I wouldn't built it directly in konserve and rather try to implement a transaction protocol on top. This is not trivial though and it will have different performance penalties. Regarding underlying transactional guarantees, this is sadly something which is very backend specific and including multi-key transactions in the interface wouldn't work for riak, couchdb, filesystem etc. It can be implemented but it has performance penalties (requires a transactor). 

I would love to have a Firebase backend, feel free to share it early on :). I think you are suggesting to make append a protocol and allow more efficient backend implementations? This is definitely doable, I have to think about it a bit. Extending `go-locked` will be very tricky though, because then you can have dead-locks depending on your transaction scheme and you will definitely suffer performance penalties depending on the groups of keys you transact over. Basically you have to implement a semaphore system for the store. I am more keen on the persistent datastructures semantics, which avoids collisions and you can just updated very few places with new roots to these (ideally one).

I also pointed out the more efficient approach with hitchhiker trees https://github.com/replikativ/konserve/issues/8, maybe this solves the problem on a higher level for you, not sure though...

**Comment by lsenta** [comment]
Thanks for the thoughtful responses, -assoc-in would be great :)

For the multi-update:
To be honest I assume the library provides consistency and atomicity if and only if there's a single instance of the store alive at a time. So the transaction can be dealt with locally (through the `go-locked`) and a "foreach k, v: (assoc k, v)" can be provided as a default. This would work for any backend.

**Comment by lsenta** [comment]
My bad, I used atomic as in "other threads see a single operation". In the presence of failures looping doesn't hold.

**Comment by whilo** [comment]
Ok, feel free to think in this direction though and build something on top if you like. Building reasonable cross-platform building blocks for databases is the main purpose of the replikativ stack (besides being useful early on). I can readd the -assoc-in protocol for you, if you like. 

Do you have the firebase code somewhere? I think many ClojureScript frontend people would love to have it. I talked to a guy at :clojureD and he was also using it with operational transforms I think. In a sense it is very strong proprietary competition for the functionality of replikativ and I am curious how people use it. I have not used it myself yet.

**Comment by whilo** [comment]
I have added an `-assoc-in` protocol. You need to be aware that you need to deal with nesting as well, but maybe you can still short circuit if you just assoc the first key new (so you can safe the roundtrip then). Is this sufficient for you?

**Comment by lsenta** [comment]
It is, thanks a lot for providing this patch @whilo!
Let's close the ticket, I'll implement an unsafe multi-key update as an extension.

Firebase is structured as a giant json object. Which fits 1:1 the Clojure map model and konserve's. So nesting is super straightforward.

I've started to split my prototype into open source libraries, be warned it's not in a usable state yet, but if you have some feedback on the implem, I'll take it :)
https://github.com/lsenta/konserve-firebase

I'll create a PR once the library is ready to be used.

**Comment by whilo** [comment]
This is great news! The code looks fine as far as I can judge on a quick look. 

One issue is the custom firebase encoding. You probably want to use it and you properly throw if somebody tries to store an unsupported type as you do, but this part needs to be documented and a test should make sure that the error propagates properly through konserve, as it will probably happen often unintendedly. This backend will not support a general EDN serializer then, but this is fine if firebase's semantics want this. Maybe you can describe the tradeoffs here? On the other hand transit supports plain JSON as a serialization format, so maybe you can even just use the transit serializer for konserve?

I am really looking forward to adding it to the README, ping me when the tests pass and I'll do so.
